### PR TITLE
Fix pthread

### PR DESCRIPTION
**Pull request by qianxichen233**
This PR fixed issue https://github.com/Lind-Project/lind-wasm/issues/32, and did some clean up and re-structure to remove some warnings and make the code more readable.

The issue was due to incorrectly handle of unwind data storage. Previously all the unwind data was stored directly at the top of the memory (offset 0). It turns out this is the place where global variables store and storing unwind data here would overwrite the data here. This is fixed by adding an option to wasm-ld to let it output the stack_low global, which represents the top of the stack. This address can then be used to store the unwind data and should not have any conflict with other data.

There is a new issue I encountered while fixing pthread. I added a chained thread test to the test list, which will currently fail because it seems like when recursively creating threads, glibc will try to use a slightly different routine to handle, which calls mmap at some place. The mmap causes the recursive thread creation fail because vmmap PR(https://github.com/Lind-Project/lind-wasm/pull/56) hasn't been merged into main yet, so mmap is still a broken feature in this branch. Hopefully once we get mmap finalized and merged into main, this issue will go away.

### CODE REVIEW

File: src/wasmtime/src/commands/run.rs [hunk 1]
```diff
@@ -207,10 +207,10 @@ impl RunCommand {
                 // exit the cage
                 lind_syscall_api(
                     1,
-                    30 as u32,
+                    30,
```
Comment (rennergade) [inline src/wasmtime/src/commands/run.rs:210]: if we're calling syscalls directly like this this should be at least a comment saying what syscall but more appropriately we should have some defines here

File: src/wasmtime/crates/lind-multi-process/src/lib.rs [hunk 2]
```diff
@@ -21,6 +21,8 @@ const ASYNCIFY_STOP_UNWIND: &str = "asyncify_stop_unwind";
 const ASYNCIFY_START_REWIND: &str = "asyncify_start_rewind";
 const ASYNCIFY_STOP_REWIND: &str = "asyncify_stop_rewind";
 
+const LIND_FS_ROOT: &str = "/home/lind-wasm/src/RawPOSIX/tmp";
```
Comment (rennergade) [inline src/wasmtime/crates/lind-multi-process/src/lib.rs:24]: should root be tmp here?

Comment (qianxichen233) [reply src/wasmtime/crates/lind-multi-process/src/lib.rs:24]: This is the root set in RawPOSIX
https://github.com/Lind-Project/lind-wasm/blob/57876e5c51e78c9285fa6842d5e1efe5f15c745f/src/RawPOSIX/src/safeposix/syscalls/fs_calls.rs#L31
I just used the same path here

File: src/wasmtime/crates/lind-multi-process/src/lib.rs [hunk 3]
```diff
@@ -231,97 +212,50 @@ impl<T: Clone + Send + 'static + std::marker::Sync, U: Clone + Send + 'static +
...
+        // -------------------------- <----- unwind_data_end (user's current stack pointer)
+        // |         .....          | ^
+        // |         .....          | |
+        // |       stack data       | | user's stack grow direction
+        // |         .....          | |
+        // |         .....          | |
+        // -------------------------- <----- stack high
+        unsafe {
+            // 16 because it is the size of two u64
+            *(unwind_data_start_sys as *mut u64) = unwind_data_start_usr + 16;
```
Comment (rennergade) [inline src/wasmtime/crates/lind-multi-process/src/lib.rs:248]: again defines are better than magic numbers

File: src/wasmtime/crates/lind-multi-process/src/lib.rs [hunk 4]
```diff
@@ -331,14 +265,14 @@ impl<T: Clone + Send + 'static + std::marker::Sync, U: Clone + Send + 'static +
...
-            println!("running out of cageid!");
+            panic!("running out of cageid!");
         }
         let child_cageid = child_cageid.unwrap();
 
         // calling fork in rawposix to fork the cage
         lind_syscall_api(
             self.pid as u64,
-            68 as u32, // fork syscall
+            68, // fork syscall
```
Comment (rennergade) [inline src/wasmtime/crates/lind-multi-process/src/lib.rs:275]: also better as a define

File: src/wasmtime/crates/lind-multi-process/src/lib.rs [hunk 5]
```diff
@@ -468,7 +399,7 @@ impl<T: Clone + Send + 'static + std::marker::Sync, U: Clone + Send + 'static +
                             // exit the cage with the exit code
                             lind_syscall_api(
                                 child_cageid,
-                                30 as u32,
+                                30,
```
Comment (rennergade) [inline src/wasmtime/crates/lind-multi-process/src/lib.rs:402]: same

File: src/wasmtime/crates/lind-multi-process/src/lib.rs [hunk 6]
```diff
@@ -526,120 +457,49 @@ impl<T: Clone + Send + 'static + std::marker::Sync, U: Clone + Send + 'static +
...
-        else {
-            println!("asyncify_start_unwind export not found");
-            return Ok(-1);
+        let asyncify_start_unwind_func = caller.get_asyncify_start_unwind().unwrap();
+
+        // store the parameter at the top of the stack
+        // reference comments in fork_call
+        unsafe {
+            // 16 because it is the size of two u64
+            *(parent_unwind_data_start_sys as *mut u64) = parent_unwind_data_start_usr + 16;
```
Comment (rennergade) [inline src/wasmtime/crates/lind-multi-process/src/lib.rs:478]: magic number

File: src/wasmtime/crates/lind-multi-process/src/lib.rs (hunk 2 above)
Comment (Yaxuan-w) [reply src/wasmtime/crates/lind-multi-process/src/lib.rs:24]: im curious that why we need using lind fs in wasmtime..?

Comment (qianxichen233) [reply src/wasmtime/crates/lind-multi-process/src/lib.rs:24]: it is used in execve. The path passed in here should be within lind filesystem

File: src/wasmtime/crates/lind-multi-process/src/lib.rs [hunk 7]
```diff
@@ -231,97 +212,50 @@ impl<T: Clone + Send + 'static + std::marker::Sync, U: Clone + Send + 'static +
...
-            return Ok(-1);
+        let asyncify_start_unwind_func = caller.get_asyncify_start_unwind().unwrap();
+
+        // store the parameter at the top of the stack
+        // we need to tell two parameters to Asyncify:
+        // unwind_data_start: the start address to store the unwind data.
+        // unwind_data_end: the end address of the avaliable space Asyncify could work with
+        // These two parameters are usually stored on the top of the unwind data.
+        // Below is a graph describing the entire user's stack layout
+        // -------------------------- <----- _usr (stack low)
```
Comment (rennergade) [inline src/wasmtime/crates/lind-multi-process/src/lib.rs:231]: i like this diagram a lot as we discussed, its still a bit confusing that unwind_data_start and end are pointers stored at that address, perhaps we can make that a bit more clear by not depicting them in large regions

**Comment by rennergade** [comment]
This should be ready to merge with the minor change requests

**Review by rennergade — APPROVED** [review]
Looks good. Think this is ready to merge.

**Pull request merged**